### PR TITLE
Introduce get_parent_run fluent API

### PR DESCRIPTION
**Pull request by annzhang-db**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

- Add get_parent_run to mlflow client API
- Add get_parent_run fluent API

## How is this patch tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Introduce `mlflow.get_parent_run()` fluent API

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by prithvikannan — COMMENTED** [review]
This looks great @annzhang-db . Can we also extend this to the mlflow client APIs (see `mlflow/tracking/client.py`)? Similar to the way we call `MlflowClient().get_run(run_id)`, it would be helpful to call `MlflowClient().get_parent_run(run_id)`.

File: mlflow/tracking/fluent.py [hunk 1]
```diff
@@ -518,6 +518,43 @@ def get_run(run_id: str) -> Run:
     return MlflowClient().get_run(run_id)
 
 
+def get_parent_run(run_id: str) -> Optional[Run]:
+    """
+    Gets the parent run for the given run id.
```
Comment (sunishsheth2009) [inline mlflow/tracking/fluent.py:523]: Gets the parent run for the given run id if one exists

**Review by sunishsheth2009 — APPROVED** [review]
Looks good to me :) Awesome work

File: mlflow/tracking/fluent.py [hunk 2]
```diff
@@ -518,6 +518,43 @@ def get_run(run_id: str) -> Run:
...
+        :caption: Output
+
+        child_run_id: 7d175204675e40328e46d9a6a5a7ee6a
+        parent_run_id: 8979459433a24a52ab3be87a229a9cdf
+    """
+    child_run = MlflowClient().get_run(run_id)
+    parent_run_id = child_run.data.tags.get(MLFLOW_PARENT_RUN_ID)
+    if parent_run_id is None:
+        return None
+    return MlflowClient().get_run(parent_run_id)
```
Comment (harupy) [inline mlflow/tracking/fluent.py:555]:
```suggestion
    client = MlflowClient()
    child_run = client.get_run(run_id)
    parent_run_id = child_run.data.tags.get(MLFLOW_PARENT_RUN_ID)
    if parent_run_id is None:
        return None
    return client.get_run(parent_run_id)
```

Can we reuse the client?

File: tests/tracking/fluent/test_fluent.py [hunk 3]
```diff
@@ -1253,3 +1253,17 @@ def test_set_experiment_tags():
...
+    mlflow.log_param("a", 1)
+    mlflow.log_metric("b", 2.0)
+    child_run_id = mlflow.start_run(nested=True).info.run_id
+    mlflow.end_run()
+    mlflow.end_run()
+
+    parent_run = mlflow.get_parent_run(child_run_id)
+    assert parent_run.info.run_id == parent_run_id
+    assert parent_run.data.metrics == {"b": 2.0}
+    assert parent_run.data.params == {"a": "1"}
```
Comment (harupy) [inline tests/tracking/fluent/test_fluent.py:1269]: Can we also add a test where the parent run doesn't exist?

**Review by harupy — APPROVED** [review]
Left comments, otherwise LGTM!

File: tests/tracking/fluent/test_fluent.py [hunk 4]
```diff
@@ -1253,3 +1253,17 @@ def test_set_experiment_tags():
...
         assert str(exact_expected_tags[tag_key]) == tag_value
+
+
+def test_get_parent_run():
+    parent_run_id = mlflow.start_run().info.run_id
+    mlflow.log_param("a", 1)
+    mlflow.log_metric("b", 2.0)
+    child_run_id = mlflow.start_run(nested=True).info.run_id
+    mlflow.end_run()
+    mlflow.end_run()
```
Comment (harupy) [inline tests/tracking/fluent/test_fluent.py:1264]:
```suggestion
    with mlflow.start_run() as parent:
        mlflow.log_param("a", 1)
        mlflow.log_metric("b", 2.0)
        with mlflow.start_run(nested=True) as child:
```

Can we use `with mlflow.start_run()`?

**Review by WeichenXu123 — APPROVED** [review]
LGTM once @harupy 's comments are addressed

File: mlflow/tracking/client.py [hunk 5]
```diff
@@ -157,6 +160,42 @@ def get_run(self, run_id: str) -> Run:
...
+            :caption: Example
+
+            import mlflow
+
+            # Create nested runs
+            with mlflow.start_run():
+                with mlflow.start_run(nested=True) as child_run:
+                    child_run_id = child_run.info.run_id
+
+            parent_run = mlflow.get_parent_run(child_run_id)
```
Comment (prithvikannan) [inline mlflow/tracking/client.py:182]: can we update this to demonstrate the MlflowClient approach? something like
```
parent_run = MlflowClient().get_parent_run
```

File: mlflow/tracking/fluent.py [hunk 6]
```diff
@@ -518,6 +518,39 @@ def get_run(run_id: str) -> Run:
...
+        print("child_run_id: {}".format(child_run_id))
+        print("parent_run_id: {}".format(parent_run.info.run_id))
+
+    .. code-block:: text
+        :caption: Output
+
+        child_run_id: 7d175204675e40328e46d9a6a5a7ee6a
+        parent_run_id: 8979459433a24a52ab3be87a229a9cdf
+    """
+    return MlflowClient().get_parent_run(run_id)
```
Comment (prithvikannan) [inline mlflow/tracking/fluent.py:551]: neat!

File: mlflow/tracking/client.py [hunk 7]
```diff
@@ -157,6 +160,42 @@ def get_run(self, run_id: str) -> Run:
         """
         return self._tracking_client.get_run(run_id)
 
+    def get_parent_run(self, run_id: str) -> Optional[Run]:
```
Comment (prithvikannan) [inline mlflow/tracking/client.py:163]: can we add a test to `tests/tracking/test_client.py`? as currently written the client and fluent APIs are the same, but would be good to keep this test coverage anyway

**Review by prithvikannan — APPROVED** [review]
Awesome work @annzhang-db ! Just one small nit comment and a test

**Pull request merged**